packages/query/src/bitcoin/ordinals/inscriptions-infinite.query.ts

```diff
@@ -0,0 +1,161 @@
+import { useCallback } from 'react';
+
+import { HDKey } from '@scure/bip32';
+import { useInfiniteQuery } from '@tanstack/react-query';
+
+import { getTaprootAddress } from '@leather.io/bitcoin';
+import { createNumArrayOfRange } from '@leather.io/utils';
+
+import { useLeatherNetwork } from '../../leather-query-provider';
+import { QueryPrefixes } from '../../query-prefixes';
+import { useBestInSlotApiRateLimiter } from '../../rate-limiter/best-in-slot-limiter';
+import { useBitcoinClient } from '../clients/bitcoin-client';
+
+const stopSearchAfterNumberAddressesWithoutOrdinals = 5;
+const addressesSimultaneousFetchLimit = 5;
+
+const inscriptionsLazyLoadLimit = 20;
+
+export function useBestInSlotGetInscriptionsInfiniteQuery({
+  taprootKeychain,
+  nativeSegwitAddress,
+}: {
+  nativeSegwitAddress: string;
+  taprootKeychain: HDKey | undefined;
+}) {
+  const network = useLeatherNetwork();
+  const limiter = useBestInSlotApiRateLimiter();
+  const client = useBitcoinClient();
+  const getTaprootAddressData = useCallback(
+    (fromIndex: number, toIndex: number) => {
+      return createNumArrayOfRange(fromIndex, toIndex - 1).reduce(
+        (acc: Record<string, number>, i: number) => {
+          const address = getTaprootAddress({
+            index: i,
+            keychain: taprootKeychain,
+            network: network.chain.bitcoin.bitcoinNetwork,
+          });
+          acc[address] = i;
+          return acc;
+        },
+        {}
+      );
+    },
+    [taprootKeychain, network.chain.bitcoin.bitcoinNetwork]
+  );
+
+  const query = useInfiniteQuery({
+    queryKey: [QueryPrefixes.GetInscriptions, 'bis', nativeSegwitAddress, network.id],
+    async queryFn({ pageParam, signal }) {
+      const responsesArr = [];
+      let fromIndex = pageParam.fromIndex;
+      let addressesWithoutOrdinalsNum = pageParam.addressesWithoutOrdinalsNum;
+      let addressesMap = getTaprootAddressData(
+        fromIndex,
+        fromIndex + addressesSimultaneousFetchLimit
+      );
+
+      let addressesData = getTaprootAddressData(
+        fromIndex,
+        fromIndex + addressesSimultaneousFetchLimit
+      );
+
+      let offset = pageParam.offset;
+      // Loop through addresses until we reach the limit, or until we find an address with many inscriptions
```

Comment:
Maybe this could be extracted into a well named standalone function so we wouldn't need the comment?

Reply:
this whole logic could be refactored/rethink, I'd prefer to do it overall later rather than partially now